koku/providers/test/aws/tests_aws_provider.py

```diff
@@ -209,3 +209,79 @@ def test_check_cost_report_access_fail(self, mock_boto3):
         access_exists = _check_cost_report_access(access_key_id, secret_access_key,
                                                   session_token)
         self.assertFalse(access_exists)
+
+    @mock_sns
+    @mock_sts
+    @mock_s3
+    @patch('providers.aws.aws_provider._check_org_access')
+    @patch('providers.aws.aws_provider._check_cost_report_access')
+    def test_get_sns_topics(self, check_org_access, check_cost_report_access):
+        """Verify when the bucket is configured for notifications."""
+        check_org_access.return_value = True
+        check_cost_report_access.return_value = True
+
+        iam_arn = 'arn:aws:s3:::my_s3_bucket'
+        bucket_name = 'my_s3_bucket'
+        access_key_id, secret_access_key, session_token = _get_sts_access(
+            iam_arn)
+        s3_resource = boto3.resource(
+            's3',
+            aws_access_key_id=access_key_id,
+            aws_secret_access_key=secret_access_key,
+            aws_session_token=session_token,
+        )
+
+        # Create bucket
+        s3_resource.create_bucket(Bucket=bucket_name)
+
+        # Create SNS Topic
+        conn = boto3.client('sns', region_name='us-east-1')
+        response = conn.create_topic(Name='CostUsageNotificationDemo')
+        topic = response['TopicArn']
+
+        # Setup Notification on bucket
+        config = {'TopicConfigurations': [{'TopicArn': topic, 'Events': ['s3:ObjectCreated:*']}]}
+        boto3.client('s3').put_bucket_notification_configuration(Bucket=bucket_name,
+                                                                 NotificationConfiguration=config)
+
+        provider_interface = AWSProvider()
+
+        try:
+            provider_interface.cost_usage_source_is_reachable(iam_arn, bucket_name)
+        except Exception:
+            self.fail('Unexpected Error')
```

Comment:
I think this try/except should be an assertion, instead.

Reply:
I'm not expecting any exceptions to be thrown.  If one is thrown then it will fail the test.  I'm not sure how that's any different from an assertion.  Even if we went with an assertion what would you assert?

Reply:
The test will fail when an exception is thrown, regardless. So, it's not necessary to explicitly call `self.fail()`. Simply not catching the exception will also be a failure - the try/catch is unnecessary.

I would expect a function that checks availability to return a boolean telling me whether or not the check(s) succeeded or not. So, then I'd expect the unit test to be an `self.assertTrue`

Reply:
Yeah thats true it would fail regardless, the try/except is just there to have some failure message.  In this case 'Unexpected Error'

So the purpose of the tests is to establish the conditions that calling `cost_usage_source_is_reachable` can encounter.  Those are 1) S3 bucket has notifications turned on 2) S3 bucket has notifications turned off.  In both of these situations I want to make sure that no exceptions are thrown..  As far as verifying other behavior change all we are doing is logging a message, so there's no systemic change to verify other than the fact that no exceptions are thrown for S3 event notifications.

The `cost_usage_source_is_reachable` regrettably returns an exception when things are bad and does nothing when things are good.  This was a carry over from the original implementation in the serializer code, but anyway..

Do you see anything wrong with the actual implementation?

Reply:
Hmmm...  I'd consider it a bug that the function doesn't return anything and relies on throwing exceptions when things are wrong. That's not a great experience. But, it's also not the code this PR is changing.  So, no, _technically_ there's nothing wrong with your specific PR.

Reply:
Yup, 100% agree with that.  Before isolating the provider-specific code in koku, the provider serializer was essentially raising assertions as the filter checks.  Instead of changing the behavior I just carried that forward since I wanted it to work the same way after moving the logic.  It works, and it's usage is very isolated so I didn't see any real reason to spend a lot of time changing it..